### ISSUE TITLE
Golang ./... wildcard path syntax regression

### ISSUE THREAD

**Issue by malcolmrebughini** [nushell/nushell#13303]
### Describe the bug

There was a bug before described [here](https://github.com/nushell/nushell/issues/7015) that has started happening again. The regression started in 0.95.0, was working in 0.94.2

### How to reproduce

Run `go test ./...` or any go command that uses that syntax

### Expected behavior

The commands run correctly

### Screenshots

![1720200661-wayshot](https://github.com/nushell/nushell/assets/9681621/d2108645-dc51-4801-be4a-8c644013835a)

### Configuration

| key                | value                                                            |
| ------------------ | ---------------------------------------------------------------- |
| version            | 0.95.0                                                           |
| major              | 0                                                                |
| minor              | 95                                                               |
| patch              | 0                                                                |
| branch             | makepkg                                                          |
| commit_hash        | f93c6680bd5ae158190556eb1afc924b0290b52f                         |
| build_os           | linux-x86_64                                                     |
| build_target       | x86_64-unknown-linux-gnu                                         |
| rust_version       | rustc 1.79.0 (129f3b996 2024-06-10) (Arch Linux rust 1:1.79.0-1) |
| cargo_version      | cargo 1.79.0                                                     |
| build_time         | 2024-06-25 20:17:10 +00:00                                       |
| build_rust_channel | release                                                          |
| allocator          | mimalloc                                                         |
| features           | default, sqlite, system-clipboard, trash                         |
| installed_plugins  |                                                                  |

**Comment by fdncred** [comment]
Apparently, whomever fixed it previously didn't add a test for it. We'll need to be sure that happens this time around if someone fixes it.

**Comment by IanManske** [comment]
In this case, I think you need to add quotes (single or double) around the path. I.e., `go test './...'`.

Bare words in argument position to external commands have path and glob expansion performed on them before they are passed to the external command. This includes n-dots expansion. E.g.,, `...` is the same as `../..` and `....` is the same as `../../..`. I'm not sure when or why n-dots support was added, but I believe it is something that was always intended to work. Previous nushell versions had incorrect implementations.

**Comment by devyn** [comment]
Yeah, it's a little like using a command that assigns a special meaning to tilde (~). There's no reason you can't, but it looks a lot like a filepath and so you can't assume that we won't do anything to it.

One option you can use if you don't want Nushell to ever do path expansion on Go commands is to define an `extern` that declares all of the arguments as strings:

```
extern go [...args: string]
```

This will cause Nushell to not do what it normally does for external commands w.r.t. path expansion and option handling. YMMV, but you can see if you like that better.

**Comment by adamkpickering** [comment]
I just upgraded my nushell installation from 0.94.2 and ran into this. It appears not just with `go` subcommands, but with other tools in the golang ecosystem that use the same wildcard syntax. Examples are `deadcode` and `golangci-lint`.

I am dismayed that the `...`, `....` etc shortcuts are a thing - as someone who primarily develops in golang, it makes my life harder. As previous commenters noted, I either have to encapsulate every `./...` in quotes, or I have to add an `extern` statement to my nushell config for every command that uses this syntax. Neither of these feels like a real solution. Golang is a popular language and a lot of people are going to be tripped up by this.

I propose adding a property to nushell config that makes it possible to turn the `...`, `....` expansion on or off. IMO would be better to have it off by default, since that would be more natural for those coming from shells without the `...` expansion (i.e. bash).

**Comment by koalazub** [comment]
> Apparently, whomever fixed it previously didn't add a test for it. We'll need to be sure that happens this time around if someone fixes it.

Maybe it's a good idea to add a warning or something to state a workaround until triage is complete and a fix is out? I originally thought it was a golang env issue before looking at the shell

**Comment by ysthakur** [comment]
> I have to add an `extern` statement to my nushell config for every command that uses this syntax

@adamkpickering To clarify: @devyn's solution [above](https://github.com/nushell/nushell/issues/13303#issuecomment-2226817630) should work even for subcommands of `go`. You don't need a separate definition for every subcommand, because `go test ./...` is considered calling the `go` command with arguments `test` and `./...`. So you only need to add that one line to your config.

Here you can see `go test ./...` works with just that one definition:
![image](https://github.com/user-attachments/assets/f713a4fe-480c-4f17-bfad-8f2cd970287e)

**Comment by oderwat** [comment]
@ysthakur this would be a solution if one did not need to specify paths in a lot of go commands. This becomes a problem quickly. Many projects have multiple apps in a `cmd` directory. This means you "go run/build/install/vet ./cmd/something/" quite a lot. I also have different tests in a `pkg`directory with 10+ subdirectories. After adding the alias you do not get any path completion. And using `./...` is especially common with exactly these types of projects.

One could maybe create a "stub" that still lists the files, but this would interfere with carapace or other completers and makes it a very complicated setup.

I really do not understand why nushell opted to implement something like `./...` as parent of parent and did not add a feature flag. You already can do `./../..` for that and this is what 100% of the shell users expect. I would also find ok to have "special syntax" in the backticks.

P.S.: I actually wonder if anybody even uses `./...` or `./........`.

**Comment by ysthakur** [comment]
> After adding the alias you do not get any path completion. And using ./... is especially common with exactly these types of projects.
@oderwat Hmm, do you have any other completions set up? Are you using Carapace (it doesn't give me any completions when I have it enabled, both with and without the extern)? I was able to get path completions after adding `extern go`.

I don't think putting n-dots expansion behind a config option would mean that a script with `foo ...` would have different behavior for different people, based on their configs, even though the script itself remains the same. So I don't think that's the best option here.

That said, your suggestion about not expanding with the `./` prefix is actually something that's been discussed before and has support, it's just that no one's implemented it yet. It'd be great if someone could put up a PR for that.

**Comment by ysthakur** [comment]
If someone wants to work on this, I think it was decided that we wanted to not perform n-dots expansion on strings beginning with `./`. That should fix `./...` specifically.

(making this a separate comment so it's not lost in my previous comment)

**Comment by oderwat** [comment]
I use carapace. With the extern I am missing all but the first completion, but the `go test ./...` works.

![image](https://github.com/user-attachments/assets/8b7c6954-1660-451d-998d-cd9afdb42994)

Without the extern I get the full completions:

![image](https://github.com/user-attachments/assets/45a3b169-4fc0-412a-b4bc-e4c313d34a9b)

**Comment by ysthakur** [comment]
@oderwat I got the same results both with and without the extern. This looks like a Carapace-specific thing. We should probably fall back to file completion if no completions are provided by the external completer, but anyway, off-topic for this issue.

**Comment by oderwat** [comment]
@ysthakur is it maybe about the order? I have carapace before extern. And that file completion fallback is interesting for me too, I just added #14582 so this is not going totally of track :)
